Fix color contrast issues in ORA

cahrens opened the pull request:
## [TNL-5104](https://openedx.atlassian.net/browse/TNL-5104)

### Description

Fix color contrast issues in ORA. Also has a change to address flaky a11y tests (TNL-5905)

### Sandbox
- [x] https://a11y.sandbox.edx.org/courses/course-v1:edx+ORA203+course/courseware/a4dfec19cf9b4a6fb5b18be6ccd9cecc/4dbd99f691064633a9be0daadb805ce6/

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @staubina 
- [x] Code review: @andy-armstrong 
- [ ] Doc Review: @catong FYI
- [x] UX review: @chris-mike 
- [x] Accessibility review: @cptvitamin FYI
- [x] Product review: @sstack22 

### Post-review
- [x] Rebase and squash commits

cahrens commented on openassessment/templates/openassessmentblock/oa_base.html:
```diff
@@ -21,12 +21,14 @@ <h2 class="openassessment__title problem__header">{% trans title %}</h2>
-                                <span class="wrapper--copy">
-                                    <span class="step__label">{% trans assessment.title %}</span>
-                                </span>
+                                <button class="ui-slidable" aria-expanded="false" disabled>
```
This is a "dummy step" shown while the actual steps (self assessment, peer, final grade, etc.) are loading. When they finish loading, each step replaces its dummy section with its actual content. The actual content has a button for the step name, and it expands/collapses to show the content.

When I removed the color-contrast ignore, I was getting failures for steps when they are still showing this "dummy" content. The name of the step was being displayed in a pale font ("disabled looking") that doesn't meet a11y guidelines. In order to work around this and make the dummy content more consistent with the content that will replace it, I am making this a button and setting it to disabled. The font will be the regular dark gray (accessible) font, but because the button is disabled, the browser changes it to a reasonable disabled color. Before and after pictures are below (before had lighter step label text).
<img width="860" alt="before" src="https://cloud.githubusercontent.com/assets/484484/20153916/054ef01c-a693-11e6-986b-1852742925af.png">

<img width="846" alt="after" src="https://cloud.githubusercontent.com/assets/484484/20153922/0bd05fa2-a693-11e6-9597-b66c8be911b8.png">

cahrens commented on openassessment/xblock/static/sass/oa/elements/_staff.scss:
```diff
@@ -29,7 +29,7 @@
   .ui-staff__title {
     @extend %t-heading;
-    color: $copy-staff-color;
+    color: $heading-staff-color;
```
The staff area had a lot of gray text that wasn't accessible. It's now pretty much all white-- a bit stark looking, but works.

Before:
![image](https://cloud.githubusercontent.com/assets/484484/20154025/86c73b0e-a693-11e6-9f72-d7099563cdaa.png)

After:
![image](https://cloud.githubusercontent.com/assets/484484/20154052/a57e7378-a693-11e6-8cdd-73eb8ec30207.png)

cahrens commented on openassessment/xblock/static/sass/oa/elements/_staff.scss:
```diff
@@ -91,14 +92,10 @@
   .openassessment_student_info_form {
     margin-bottom: ($baseline-v/2);
 
-    .list--actions__item {
```
These elements are no longer wrapped in lists-- that was removed some time back, but the sass did not get updated. I wanted to change the "Submit" button above to $btn--primary, and thus found that this sass was not properly overriding.

cahrens commented on openassessment/xblock/static/sass/oa/elements/_staff.scss:
```diff
@@ -63,6 +63,7 @@
     .wrapper--content {
       padding: $baseline-v ($baseline-h/2);
       background-color: $bg-content;
+      color: $copy-color;
```
A lot of the content in the staff tools was too pale. I used the accessible $copy-color that is used in the student assessments.

Before:
![image](https://cloud.githubusercontent.com/assets/484484/20154174/22a0b014-a694-11e6-8f8e-e9c9049cef97.png)

After:
![image](https://cloud.githubusercontent.com/assets/484484/20154192/3142a474-a694-11e6-98c6-605a03c10e77.png)

cahrens commented on openassessment/xblock/static/sass/oa/utilities/_developer.scss:
```diff
@@ -64,7 +60,7 @@ $link-hover: $edx-blue-l1 !default; // from our Pattern Library http://ux.edx.or
       &:hover {
         color: white;
-        background-color: $link-hover;
+        background-color: $action-primary-color;
```
This is the pattern library blue, defined in variables.

cahrens commented on openassessment/xblock/static/sass/oa/utilities/_developer.scss:
```diff
@@ -163,7 +159,7 @@ $link-hover: $edx-blue-l1 !default; // from our Pattern Library http://ux.edx.or
   .student-training__display__title {
     @extend %t-heading;
     margin-bottom: ($baseline-v/2);
-    color: $heading-secondary-color;
```
We are losing the subtlety of different heading levels having slightly different colors, but now they are at least accessible.

cahrens commented on openassessment/xblock/static/sass/oa/views/_oa-base.scss:
```diff
@@ -286,7 +286,7 @@
 
       .step__counter:before, .step__label, .step__deadline {
-        color: $color-unavailable;
+        color: $copy-color;
```
Step counters before were often a very pale color. Make them the regular, accessible text color.

Before:
![image](https://cloud.githubusercontent.com/assets/484484/20154427/3c3d3050-a695-11e6-8dd0-274a7fa54a64.png)

After:
![image](https://cloud.githubusercontent.com/assets/484484/20154437/461dc1e8-a695-11e6-8ea3-72a1f2a6b175.png)

cahrens commented on openassessment/xblock/static/sass/themes/edx/_utilities.scss:
```diff
@@ -10,47 +10,11 @@
 // --------------------
 $edx-blue: rgb(58, 162, 224);
 $edx-blue-l1: tint($edx-blue,20%);
-$edx-blue-l2: tint($edx-blue,40%);
```
Deleting a lot of variables that aren't used (or have other values in _variables.scss).

When the code was originally written, there was an attempt to have xblock-general colors and then edX overrides. No other repo is using this though, and at this point trying to keep them separate only adds confusion, duplication, etc. I did a lot of slash and burn in this PR.

andy-armstrong replied:
💯

cahrens commented on openassessment/xblock/static/sass/xb/elements/_controls.scss:
```diff
@@ -26,40 +26,6 @@
   }
 }
 
-// links - meta
```
Unused stuff (intended to be generally useful for xblocks).

cahrens commented on openassessment/xblock/static/sass/xb/elements/_controls.scss:
```diff
@@ -146,74 +111,45 @@
-  background: $action-secondary-color;
-  border: 1px shade($action-secondary-color, 10%);
-  color: tint($action-secondary-color, 90%);
+  background: $white;
```
Secondary button is white blackground with blue border/text, reverses on hover. Disabled is a grayish also (from pattern library).

cahrens commented on openassessment/xblock/static/sass/xb/elements/_system-feedback.scss:
```diff
@@ -42,50 +42,40 @@
   }
 
-  // TYPE: confirmation
-  .message--confirmation {
```
Unused.

cahrens commented on openassessment/xblock/static/sass/xb/utilities/_variables.scss:
```diff
@@ -39,57 +39,25 @@ $gray-d3: shade($gray,60%);
 $gray-d4: shade($gray,80%);
 $gray-d5: shade($gray,90%);
 $gray-d6: shade($gray,95%);
-$gray-d7: shade($gray,99%);
```
Deleting unused variations.

cahrens commented on test/acceptance/accessibility.py:
```diff
@@ -24,16 +24,11 @@ def _configure_a11y_audit_config(self, page):
                 "#footer-edx-v3",  # Links to Facebook, Twitter, etc. Populated in production, but not here.
-                ".instructor-info-action",  # Staff Debug Info link (general for all XBlocks).
+                ".instructor-info-action",  # Staff Debug Info link (general for all XBlocks)
+                ".is--disabled", # TODO: TNL-5183
```
Buttons that are disabled only via a class still fail color contrast. We can remove this exclude when tackling TNL-5183 (disabled buttons should be truly disabled...).

cahrens commented on test/acceptance/pages.py:
```diff
@@ -356,6 +360,11 @@ def wait_for_num_completed(self, num_completed):
+        # This is an evil anti-pattern. Was experiencing sporadic failures when completing assessments after
+        # the first, and could not find anything to wait on that fixed the failures.
+        # Investigate further in TNL-5905
+        if num_completed > 0:
```
Will remove with TNL-5905.

cahrens commented on openassessment/xblock/static/sass/oa/utilities/_shame.scss:
```diff
@@ -142,7 +142,7 @@
   // actions
-  .action--submit, .action--save {
+  .action--submit, .action--save, .action--upload {
     text-shadow: none !important;
```
FYI, here you can see the constellation of different buttons:

![image](https://cloud.githubusercontent.com/assets/484484/20154768/efee092a-a696-11e6-8695-49c08081decb.png)

chris-mike commented:
Looking good, @cahrens .

On the first page, it looks like the grey boxes are not the same color (the darker one) and they should be.  (this darker grey: palette(grayscale, back)RGB: rgb(217, 217, 217) HEX: d9d9d9)
(see screenshot)

![ora-page1](https://cloud.githubusercontent.com/assets/16578372/20193692/25e38232-a75d-11e6-9acf-baab9384956f.png)

chris-mike commented:
Another thought: would it be easy to make "in progress" blue, rather than grey. Something like this: 

![screen shot 2016-11-10 at 3 59 23 pm](https://cloud.githubusercontent.com/assets/16578372/20194046/c59f55b6-a75e-11e6-95f3-3ffd9d27902f.png)

cahrens commented:
@chris-mike Thanks for taking a look and providing some feedback!

I updated the gray top border and the background of the progress boxes (when gray) to be $uxpl-gray-back-- rgb(217, 217, 217). Previously it was inconsistent, sometimes a slightly different gray.

And I made the "In Progress" color blue ($uxpl-primary-base, rgb(0, 117, 180). In your last screenshot, you had it the orangish color instead of blue-- was that intentional? In what I pushed up, it will always be blue.

![image](https://cloud.githubusercontent.com/assets/484484/20234757/29416816-a84f-11e6-88d9-689a3ba3de21.png)

cahrens commented:
@chris-mike OK, as we talked about, I made the "peer waiting" template use the is--incomplete class, which renders the warning color.

![image](https://cloud.githubusercontent.com/assets/484484/20271024/571fcbc4-aa57-11e6-9366-265759ac8fb5.png)

cahrens commented on openassessment/templates/openassessmentblock/peer/oa_peer_waiting.html:
```diff
@@ -2,7 +2,7 @@
 
 {% block list_item %}
-<li id="openassessment__peer-assessment__{{ xblock_id }}" class="openassessment__steps__step step--peer-assessment is--showing ui-slidable__container" tabindex="-1">
+<li id="openassessment__peer-assessment__{{ xblock_id }}" class="openassessment__steps__step step--peer-assessment is--incomplete is--showing ui-slidable__container" tabindex="-1">
```
Waiting for peer assessments to grade will now render with the yellowish "warning" color, which is mapped to the pre-existing is--incomplete class.

cahrens commented on test/acceptance/pages.py:
```diff
@@ -215,12 +218,30 @@ def assess(self, options_selected):
+            return True
+
+        attempts = 0
+        while not criterion_selected() and attempts < 5:
```
This retry logic is because sometimes the selection of options in the 2nd training assessment fails (the options do not get selected, and therefore the test times out because the Submit button never enables). I tried waiting on various things to ensure that the second assessment was fully loaded, but nothing helped. Retrying up to at most 5 times seems to have made this test stable on devstack (it very rarely ever failed on Jenkins... the flaky issue was primarily devstack), and it adds very little extra time to the test. However, I fully admit it is still a hack. I just don't have any other ideas at the moment.

cahrens replied:
ORA tests don't use the flaky decorator-- I'm not sure if they can. The acceptance tests have some retry logic (there is a retry decorator defined in tests.py), but the accessibility tests haven't used it.

My main concern about doing a flaky/retry thing for these tests is that the 2 particular tests that include training are EXTREMELY long and slow. There is a lot of logging in as different users and going back and forth to make it through the whole workflow. If a failure is hit in training, starting over from the beginning will add quite a bit of time. Also, since the failure typically happens at the same point (the beginning of the second training exercise), starting all over is less likely to be successful vs. re-clicking on the assessment that is already being shown.

staubina replied:
Ok, thank you for clarifying.

cahrens replied:
I'll add a comment though in the code.

sstack22 commented:
Looks good to me, though I wonder if the "Your grade" being blue, even when the grade is not available, is confusing? CC: @cahrens , @chris-mike .

cahrens commented:
@sstack22 "Your Grade" is blue because it is clickable-- the section expands to reveal additional information.

It's true that blue here is overloaded. It's being used for links, blue action buttons, "link-like things" that are being used as buttons, and "In Progress". Any concerns, @chris-mike?

staubina commented:
Minor request to add comments to the flaky test.  Other than that these changes look good to me.

andy-armstrong commented on openassessment/xblock/static/sass/oa/elements/_staff.scss:
```diff
@@ -91,14 +92,10 @@
 
-        margin-left: 0;
-      }
+      margin-left: 0;
```
This isn't your change, but this isn't RTL-compatible. It should use:

    @include margin-left(0);

andy-armstrong commented on openassessment/xblock/static/sass/oa/utilities/_developer.scss:
```diff
@@ -27,10 +27,6 @@
-$lighter-base-font-color: rgb(100,100,100);
-$link-hover: $edx-blue-l1 !default; // from our Pattern Library http://ux.edx.org/elements/colors/
-
 .wrapper--openassessment .wrapper--staff-area {
```
Nit: at some point we should remove this file since we don't have a UX/developer distinction anymore for Sass development. It might be a good time to make this change now, but I understand if that seems out-of-scope.

cahrens replied:
I'd prefer not to do it as part of this story. The question though is when would we do it.... It would be good to get rid of the edx "theme" files too.

andy-armstrong commented on openassessment/xblock/static/sass/oa/views/_oa-base.scss:
```diff
@@ -389,7 +389,31 @@
+    &.is--in-progress {
+
+      .step__counter:before {
+        color: $uxpl-primary-base;
```
Nit: I'd prefer this use a semantic variable name, so that it is clearer how to override it in a custom theme. I'm not sure if there is a natural one to use, or whether a new one would have to be introduced.

cahrens replied:
I will switch to using $action-primary-color, which I think conveys the intent here.

andy-armstrong commented on openassessment/xblock/static/sass/oa/views/_oa-base.scss:
```diff
@@ -389,7 +389,31 @@
+        }
+
+        .copy, .icon {
+          color: $white;
```
I'd also like this to be a semantic variable name, e.g. ``$copy-inverse-color``.

cahrens replied:
Done, and fixed other places where the copy (text) color was set to white.

andy-armstrong commented on openassessment/xblock/static/sass/xb/utilities/_variables.scss:
```diff
@@ -114,14 +83,29 @@ $tmg-f3: 0.125s;
+//---------------------
+// UXPL colors
+//---------------------
+$uxpl-gray-base: rgb(65, 65, 65);
```
I don't think this is possible with the current state of the Pattern Library, but it would have been great to have npm installed the library and then just use the constants from there. We need to work on making that possible.

cahrens commented:
@andy-armstrong and @staubina I have responded to your feedback and updated the sandbox (not that anything that I changed should have made a visual difference...).

chris-mike commented:
@cahrens
I have no specific issues with blue. The section titles that were grey  could stay gray as long as the continue to change on hover (there was an interaction there that I do not see on my phone). But I'm also good with what we have now. 

Btw reviewing this on my phone, there are definitely responsive issues.
